File: 0540-single-element-in-a-sorted-array/0540-single-element-in-a-sorted-array.java
class Solution {
    public int singleNonDuplicate(int[] nums) {
        //Get the size of the array
        int n =nums.length;
        //Edge Case: Only one element in the array
        if(n==1) return nums[0];
        //Edge Case: First Element is the unique one
        if (nums[0] != nums[1]) return nums[0];
        //Edge Case: Last Element is the unique one
        if (nums[n-1]!= nums[n-2]) return nums[n-1];

        int low=1, high = n-2;
        //Perform Binary Search
        while(low<=high){
            int mid =  low+(high-low)/2;
            //Check if middle element is the unique one
            if(nums[mid]!= nums[mid-1] && nums[mid]!=nums[mid+1]){
                return nums[mid];
            }
            if ((mid % 2 == 1 && nums[mid] == nums[mid - 1]) || //odd
                (mid % 2 == 0 && nums[mid] == nums[mid + 1])) {
                // Move to the right half
                low = mid + 1;
            }
            // If mid is in the right half (pairing broken earlier)
            else {
                // Move to the left half
                high = mid - 1;
            }
        }

        // Dummy return (not reachable if input is valid)
        return -1;
        }
        
    }

File: 0540-single-element-in-a-sorted-array/0540-single-element-in-a-sorted-array.py
class Solution(object):
    def singleNonDuplicate(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        n= len(nums)
        #Edge Case: Only one element in the array
        if n==1:
            return nums[0]
        #Edge Case: First element is the unique number 
        if nums[0]!=nums[1]:
            return nums[0]
        #Edge Case: Last Element is the unique number
        if nums[n-1]!=nums[n-2]:
            return nums[n-1]
        low, high = 1, n-2
        #Perform Binary Search 
        while(low<=high):
            mid = (low+high)//2
            #check if middle element is the unique one 
            if nums[mid]!=nums[mid+1] and  nums[mid]!=nums[mid-1]:
                return nums[mid]
            #if mid is in the left half pairing is valid
            if (mid %2 ==1 and nums[mid]==nums[mid-1]) or \
               (mid%2==0 and nums[mid]==nums[mid+1]):
               #Move to the right half 
               low = mid+1
            else:
                #Move to the lefr half 
                high = mid -1
        # Dummy return (not reachable if input is valid)
        return -1
        